### PR TITLE
polish: replace ported state diff with lotus actor registry

### PR DESCRIPTION
**Pull request by frrist**
### What
Remove the ported state diff and replace it with a more maintainable implementation that pulls actor method names and param parsing from lotus internals. The format of the params as json will change slightly, need to confirm this is acceptable.

Makes progress towards #602

### CODE REVIEW

File: tasks/messages/message_test.go [hunk 1]
```diff
@@ -106,13 +106,15 @@ func TestParseMessageParams(t *testing.T) {
...
+			// https://filfox.info/en/message/bafy2bzacebpiuu7tgya6yz56sfllpqc3rqbo5s5xl7353xeuavc53qlpb4sqw
 			name:        "issue-709",
 			method:      1,
 			params:      nil,
 			actorCode:   builtin3.AccountActorCodeID,
 			wantMethod:  "Constructor",
-			wantEncoded: `""`,
-			wantErr:     false,
+			wantEncoded: ``,
+			wantErr:     true,
```
Comment (frrist) [inline tasks/messages/message_test.go:117]: I introduced this mistake here: https://github.com/filecoin-project/lily/pull/714/files#diff-cbf6199b767dd5286dea0dea57200b50ead0e919d0079734bc3e0cc2bb006fc1R114.
An error is expected.

File: tasks/messages/message_test.go [hunk 2]
```diff
@@ -48,7 +48,7 @@ func TestParseMessageParams(t *testing.T) {
 			method:      2,
 			params:      mustDecodeBase64(t, "gm5maWwvMi9tdWx0aXNpZ1hFhVUBVZOiUErA4tz2x5Zz9iV/aBZBdHxVARJCEmhAYnLXQYa/1IScyah3BRgEVQE3/ahkH/xoK6Xpa0ZbDxkY/oPxiAIA"),
 			actorCode:   builtin5.InitActorCodeID,
-			wantMethod:  "InitExecParams",
+			wantMethod:  "Exec",
```
Comment (frrist) [inline tasks/messages/message_test.go:51]: The actual method name is just `Exec`: https://github.com/filecoin-project/specs-actors/blob/v6.0.1/actors/builtin/init/init_actor.go#L23

File: lens/util/repo.go [hunk 3]
```diff
@@ -194,31 +203,52 @@ func GetExecutedAndBlockMessagesForTipset(ctx context.Context, cs *store.ChainSt
...
+	params, method, err := ParseParams(m.Params, m.Method, destCode)
 	if method == "Unknown" {
 		return "", "", xerrors.Errorf("unknown method for actor type %s: %d", destCode.String(), int64(m.Method))
 	}
 	if err != nil {
-		log.Warnf("failed to parse parameters of message %s: %v", m.Cid(), err)
+		log.Warnf("failed to parse parameters of message %s: %v", m.Cid, err)
 		// this can occur when the message is not valid cbor
-		return method, "", xerrors.Errorf("failed to parse parameters of message %s: %w", m.Cid(), err)
+		return method, "", err
```
Comment (iand) [inline lens/util/repo.go:245]: Any reason to drop the xerrors context for this error?

File: lens/util/repo.go [hunk 4]
```diff
@@ -194,31 +203,52 @@ func GetExecutedAndBlockMessagesForTipset(ctx context.Context, cs *store.ChainSt
...
 		return method, "", nil
 	}
 
-	buf := bytes.NewBuffer(nil)
-	if err := fcjson.Encoder(params, buf); err != nil {
-		return "", "", xerrors.Errorf("json encode message params: %w", err)
-	}
-
-	encoded := string(bytes.ReplaceAll(bytes.ToValidUTF8(buf.Bytes(), []byte{}), []byte{0x00}, []byte{}))
+	encoded := strings.ReplaceAll(strings.ToValidUTF8(params, ""), string([]byte{0x00}), "")
```
Comment (iand) [inline lens/util/repo.go:251]: This was a hack to work around an issue in fcjson. We should verify if we need to do it any more.

Comment (iand) [reply lens/util/repo.go:251]: There is an issue that contains a message cid we could use to make a test https://github.com/filecoin-project/lily/issues/741

**Review by iand — APPROVED** [review]
Looks good. We could get a lot more confidence if we retrieved some examples of real messages and added tests for them. I don't think that would be much effort since we have the test harness.

File: lens/util/repo.go (hunk 3 above)
Comment (frrist) [reply lens/util/repo.go:245]: whoops no, reverting

File: tasks/messages/message_test.go [hunk 5]
```diff
@@ -122,20 +126,95 @@ func TestParseMessageParams(t *testing.T) {
...
 			name:        "issue-772",
 			method:      2,
 			params:      mustDecodeBase64(t, "dHJhbnNmZXI="),
 			actorCode:   builtin3.AccountActorCodeID,
 			wantMethod:  "PubkeyAddress",
-			wantEncoded: `"dHJhbnNmZXI="`,
+			wantEncoded: ``,
 			wantErr:     false,
 		},
+		// Derived from message bafy2bzaceclkxipts274qqchuqn4ri6w5cce26f7lbxljfpbsgalv4zeq4huk
```
Comment (frrist) [inline tasks/messages/message_test.go:143]: @iand I have attempted to check that https://github.com/filecoin-project/lily/issues/741 via this test. Mind giving this a second look? I don't feel very certain on the fix 😶‍🌫️

File: lens/util/repo.go [hunk 6]
```diff
@@ -205,33 +215,52 @@ func GetExecutedAndBlockMessagesForTipset(ctx context.Context, cs *store.ChainSt
...
 
+func ParseParams(params []byte, method abi.MethodNum, actCode cid.Cid) (string, string, error) {
+	m, found := ActorRegistry.Methods[actCode][method]
+	if !found {
+		return "", "", fmt.Errorf("unknown method %d for actor %s", method, actCode)
+	}
+
+	// if the actor method doesn't expect params don't parse them
+	// messages can contain unexpected params and remain valid, we need to ignore this case for parsing.
+	if m.Params == reflect.TypeOf(new(abi.EmptyValue)) {
```
Comment (iand) [inline lens/util/repo.go:226]: LGTM

**Review by iand — APPROVED** [review]
LGTM

**Comment by frrist** [comment]
Changes introduced to the message params field as a result of this PR:

| Method Name | Previous | Current |
|---------------|----------|---------|
| Propose | {"To": "f1gojwyi2jcvptacdaopoh6as2whvwv2i67j2gqbq", "Value": "3149869747391647641", "Method": 0, "Params": ""} | {"To": "f1gojwyi2jcvptacdaopoh6as2whvwv2i67j2gqbq", "Value": "3149869747391647641", "Method": 0, "Params": null} |
| TerminatedSectors | {"Terminations": [{"Sectors": {"rle": "[14184,1]", "_type": "bitfield", "elemcount": "1"}, "Deadline": 6, "Partition": 0}]} | {"Terminations": [{"Sectors": [14184, 1], "Deadline": 6, "Partition": 0}]} |
| ProveCommitSectors | {"Proof": "omitted", "SectorNumber": 287854} | {"Proof": "omitted", "SectorNumber": 287854}|
| SubmitWindowedPost | {"Proofs": [{"PoStProof": 9, "ProofBytes": "s6R286ikYpAIu9trmrrYghMIs0qRgMcazPnur25XoGkvT+a6eQzb70bGdpovFBBisetEzoWBNEx3lwFSeTbkfVxhNP+JyyddqMUzrle0jDj+xLy7EfunGt1OumWV2TCWATHtgup43Za7LecSQg1fHW1VQpu4hynobXsvRNSrIUPXuHR5BXJUsGRn1zfrFV4Jj6Te5tJ5V3UldfUOu74Z7+3PxQeUKyzqubGxNXITxboCKNds3Q3ocOdI4T3AkyqK"}], "Deadline": 35, "Partitions": [{"Index": 0, "Skipped": {"rle": "[0]", "_type": "bitfield", "elemcount": "0"}}], "ChainCommitRand": "h1MbWURLsVmZB2FSyb4Ff5iGjsyJRpfA7UWfdOTrsg4=", "ChainCommitEpoch": 1480063} | {"Proofs": [{"PoStProof": 9, "ProofBytes": "s6R286ikYpAIu9trmrrYghMIs0qRgMcazPnur25XoGkvT+a6eQzb70bGdpovFBBisetEzoWBNEx3lwFSeTbkfVxhNP+JyyddqMUzrle0jDj+xLy7EfunGt1OumWV2TCWATHtgup43Za7LecSQg1fHW1VQpu4hynobXsvRNSrIUPXuHR5BXJUsGRn1zfrFV4Jj6Te5tJ5V3UldfUOu74Z7+3PxQeUKyzqubGxNXITxboCKNds3Q3ocOdI4T3AkyqK"}], "Deadline": 35, "Partitions": [{"Index": 0, "Skipped": [0]}], "ChainCommitRand": "h1MbWURLsVmZB2FSyb4Ff5iGjsyJRpfA7UWfdOTrsg4=", "ChainCommitEpoch": 1480063} |
| Propose | {"To": "f1em4vxj6ch7ktjqzn2bm5l6mhbfblhuzqhpvzh5y", "Value": "161000000000000000000", "Method": 0, "Params": ""} | {"To": "f1em4vxj6ch7ktjqzn2bm5l6mhbfblhuzqhpvzh5y", "Value": "161000000000000000000", "Method": 0, "Params": null} |
| PreCommitSectorBatch | {"Sectors": [{"DealIDs": [], "SealProof": 8, "SealedCID": {"/": "bagboea4b5abcasvz2vj6ftxpdjkkrt4lhh36gqcdrohemdxtjnx2efkm6sfgxptr"}, "Expiration": 3029491, "SectorNumber": 70630, "SealRandEpoch": 1478430, "ReplaceCapacity": false, "ReplaceSectorNumber": 0, "ReplaceSectorDeadline": 0, "ReplaceSectorPartition": 0}]} | {"Sectors": [{"DealIDs": null, "SealProof": 8, "SealedCID": {"/": "bagboea4b5abcasvz2vj6ftxpdjkkrt4lhh36gqcdrohemdxtjnx2efkm6sfgxptr"}, "Expiration": 3029491, "SectorNumber": 70630, "SealRandEpoch": 1478430, "ReplaceCapacity": false, "ReplaceSectorNumber": 0, "ReplaceSectorDeadline": 0, "ReplaceSectorPartition": 0}]} |
| DeclareFaultRecovered | {"Recoveries": [{"Sectors": {"rle": "[10816,31,1407,1,2,23,1,4,1,1,1405,1,3,26,1,1,1,1,521,1,3,915,58,1,2,1343]", "_type": "bitfield", "elemcount": "2349"}, "Deadline": 6, "Partition": 0}]} | {"Recoveries": [{"Sectors": [10816, 31, 1407, 1, 2, 23, 1, 4, 1, 1, 1405, 1, 3, 26, 1, 1, 1, 1, 521, 1, 3, 915, 58, 1, 2, 1343], "Deadline": 6, "Partition": 0}]}|

**Comment by frrist** [comment]
I can add more methods to the above, but the key differences from what I have scraped from my database and prod are:
- Write `null` instead of `""` or `[]`
- Sectors fields no longer contain a `__type` or `elemcount` field.

**Pull request merged**